### PR TITLE
Rename impl_abstract to register_fake, part 1/2

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* #124071
* #124066
* #124065
* #124064
* __->__ #123937

This PR:
- adds a new torch.library.register_fake and deprecates
  torch.library.impl_abstract. The motivation is that we have a lot of
  confusion around the naming so we are going to align the naming with
  the actual subsystem (FakeTensor).
- renames `m.impl_abstract_pystub("fbgemm_gpu.sparse_ops")` to
  `m.has_python_registration("fbgemm_gpu.sparse_ops")`. No deprecation
  here yet; I need to test how this works with static initialization.
- Renames a bunch of internals to match (e.g. abstractimplpystub ->
  pystub)

I'm scared to rename the Python-side internal APIs (e.g.
torch._library.abstract_impl) because of torch.package concerns. I'll do
that in its own isolated PR next just in case it causes problems.

DEPRECATION NOTE: torch.library.impl_abstract was renamed to to
torch.library.register_fake. Please use register_fake. We'll delete
impl_abstract in a future version of PyTorch.

Test Plan:
- existing tests

### CODE REVIEW

File: torch/library.h [hunk 1]
```diff
@@ -611,18 +611,23 @@ class TORCH_API Library final {
...
+  /// fake impls may be found in the given Python module (pymodule).
+  /// This registers some help text that is used if the fake impl
   /// cannot be found.
   ///
   /// Args:
   /// - pymodule: the python module
   /// - context: We may include this in the error message.
-  Library& impl_abstract_pystub(const char* pymodule, const char* context = "") {
-    impl_abstract_pystub_ = {pymodule, context};
+  Library& has_python_registrations(const char* pymodule, const char* context = "") {
```
Comment (albanD) [inline torch/library.h:621]: I would expect that this would return a bool given the function name?

Comment (zou3519) [reply torch/library.h:621]: TORCH_LIBRARY(lib, m) {
  m.set_python_module("...");
}

**Review by albanD — APPROVED** [review]
ok!

**Pull request closed without merging**